diff --git a/src/logging.c b/src/logging.c
index 18d8d7d..30ead7d 100644
--- a/src/logging.c
+++ b/src/logging.c
@@ -113,6 +113,8 @@ int _mosquitto_log_vprintf(struct mosquitto *mosq, int priority, const char *fmt
 	int syslog_priority;
 	time_t now = time(NULL);
 	static time_t last_flush = 0;
+	char rfc3339_timestamp[32];
+	strftime(rfc3339_timestamp, 32, "%Y-%m-%dT%H:%M:%SZ", gmtime(&now));
 
 	if((log_priorities & priority) && log_destinations != MQTT3_LOG_NONE){
 		switch(priority){
@@ -199,7 +201,7 @@ int _mosquitto_log_vprintf(struct mosquitto *mosq, int priority, const char *fmt
 
 		if(log_destinations & MQTT3_LOG_STDOUT){
 			if(int_db.config && int_db.config->log_timestamp){
-				fprintf(stdout, "%d: %s\n", (int)now, s);
+				fprintf(stdout, "%s: %s\n", rfc3339_timestamp, s);
 			}else{
 				fprintf(stdout, "%s\n", s);
 			}
@@ -207,7 +209,7 @@ int _mosquitto_log_vprintf(struct mosquitto *mosq, int priority, const char *fmt
 		}
 		if(log_destinations & MQTT3_LOG_STDERR){
 			if(int_db.config && int_db.config->log_timestamp){
-				fprintf(stderr, "%d: %s\n", (int)now, s);
+				fprintf(stderr, "%s: %s\n", rfc3339_timestamp, s);
 			}else{
 				fprintf(stderr, "%s\n", s);
 			}
@@ -215,7 +217,7 @@ int _mosquitto_log_vprintf(struct mosquitto *mosq, int priority, const char *fmt
 		}
 		if(log_destinations & MQTT3_LOG_FILE && int_db.config->log_fptr){
 			if(int_db.config && int_db.config->log_timestamp){
-				fprintf(int_db.config->log_fptr, "%d: %s\n", (int)now, s);
+				fprintf(int_db.config->log_fptr, "%s: %s\n", rfc3339_timestamp, s);
 			}else{
 				fprintf(int_db.config->log_fptr, "%s\n", s);
 			}
@@ -240,7 +242,7 @@ int _mosquitto_log_vprintf(struct mosquitto *mosq, int priority, const char *fmt
 					_mosquitto_free(s);
 					return MOSQ_ERR_NOMEM;
 				}
-				snprintf(st, len, "%d: %s", (int)now, s);
+				snprintf(st, len, "%s: %s", rfc3339_timestamp, s);
 				mqtt3_db_messages_easy_queue(&int_db, NULL, topic, 2, strlen(st), st, 0);
 				_mosquitto_free(st);
 			}else{
